### PR TITLE
Build locally

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by isuruf** [comment]
cc @ngam

**Comment by ngam** [comment]
Let me test this locally soon, thank you 😻

**Review by ngam — CHANGES_REQUESTED** [review]

**Comment by ngam** [comment]
@isuruf two notes while we are at it:

- build-locally.py requires python3 (as you have the shebang) however if someone runs it with python2 (like I accidentally did) it will give a confusing error (I think about the f" string). Note that python2 will remain default on a lot of machines for the foreseeable future, so we may want to add a more meaningful error. But but but --- people should be moving on from python2 lol
- I am still somewhat confused about the osx situation: `RuntimeError: Need OSX_SDK_DIR env variable set. Run 'export OSX_SDK_DIR=/opt'to download the SDK automatically to '/opt/MacOSX<ver>.sdk'`. I thought you said we had this figured out?

File: build-locally.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    if ns.config.startswith("win"):
+        raise ValueError(
+            f"only Linux/macOS configs currently supported, got {ns.config}"
+        )
+    elif ns.config.startswith("osx") and platform.system() == "Darwin":
+        if "OSX_SDK_DIR" not in os.environ:
+            raise RuntimeError(
+                "Need OSX_SDK_DIR env variable set. Run 'export OSX_SDK_DIR=/opt'"
+                "to download the SDK automatically to '/opt/MacOSX<ver>.sdk'"
+            )
```
Comment (ngam) [inline build-locally.py:73]:
```suggestion
            print(f"setting "OSX_SDK_DIR" to automatically download the SDK 'MacOSX<ver>.sdk' to the staged-recipes repository root.")
            os.environ["OSX_SDK_DIR"] = "."
```

Why don't we just do this automatically as you suggested?

Comment (isuruf) [reply build-locally.py:73]: Ah, license issues I think.

Comment (ngam) [reply build-locally.py:73]: ooooooh makes sense!!

Well at least the error message here is super clear!

**Comment by ngam** [comment]
This is really good, thank you so much @isuruf!!

**Comment by ngam** [comment]
> * build-locally.py requires python3 (as you have the shebang) however if someone runs it with python2 (like I accidentally did) it will give a confusing error (I think about the f" string). Note that python2 will remain default on a lot of machines for the foreseeable future, so we may want to add a more meaningful error. But but but --- people should be moving on from python2 lol

finally: https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python

**Comment by isuruf** [comment]
Ping on this @conda-forge/staged-recipes

File: build-locally.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+#
+# This file has been generated by conda-smithy in order to build the recipe
```
Comment (xhochy) [inline build-locally.py:3]: This file here probably hasn't been directly generated by conda-smithy but copied from a conda-smithy generated repo. Should we mention this somewhere?

Comment (isuruf) [reply build-locally.py:3]: This file is an exact copy from conda-smithy (i.e. no changes have been made). I'm not sure if it's worth it to just change the comment.

**Review by xhochy — APPROVED** [review]

**Review by beckermr — COMMENTED** [review]
So it appears this pr will build all recipes in staged-recipes. However, it can be the case that recipes which are not your own are present in master when the queue backs up. Then users will be building their recipe and a bunch of other ones iiuic. Do we need to make another PR to change this?

**Comment by isuruf** [comment]
I don't think it would build all recipes.

**Comment by beckermr** [comment]
Hmm ok. I'll have to play around with it. I can't tell from looking at the pr how to build just a single recipe.

**Comment by beckermr** [comment]
Ahhhh build_steps.sh does some git munging to remove the ones on main. This may leave the users repo in a weird spot. I can't tell from the git commands but this might be worth looking into at some point.

**Comment by jakirkham** [comment]
> Ahhhh build_steps.sh does some git munging to remove the ones on main. This may leave the users repo in a weird spot. I can't tell from the git commands but this might be worth looking into at some point.

Yeah that was really intended for CI. Idk if that should be used locally (people might commit these deletions).

**Pull request merged**